Added Docker client invocation builder; ability to form pipes (resolves #1)

hannes-ucsc opened the pull request:
Resolves #1

hannes-ucsc committed:
Added Docker client invocation builder; ability to form pipes

heuermh commented on src/toil_lib/programs.py:
```diff
@@ -124,3 +125,282 @@ def _fix_permissions(base_docker_call, tool, work_dir):
+    """
+    A builder (https://en.wikipedia.org/wiki/Builder_pattern) for Popen objects representing a
+    Docker client invocation. When building a Popen object, it distinguishes between three
+    layers, the `popen` layer, the `docker` layer and the `extra` layer. The `popen` layer
```
I don't like `extra` as a layer name, because it is never referred to as such in Docker docs.  Leaving that off is perhaps not clear enough?
```python
(Docker()
  .docker_command('run') # invoke `docker run`
  .docker_params('ubuntu', interactive=True) # pass `--interactive` and run `ubuntu` image
  .docker_params(rm=True)  # docker() is cumulative, also pass `--rm`
  .command('foo') # invoke `foo` command inside container
  .params('/', bar=42) # pass `--bar=42 /` to foo
  .popen_params(cwd='/foo/bar') # pass `cwd='/foo/bar'` to the Popen constructor
  .work_dir('/foo/bar')  # same
  .build_popen_call())
```

hannes-ucsc replied:
I agree `extra` is bad and I like just leaving it off. So 

extra_command => command
extra_params =>params

but there also the `extra(command,*args,**opts)` shortcut. What do we name that?

hannes-ucsc committed:
More tweaks

- revised handling of boolean options (foo=False removes previous occurrences of --foo rather than add --foo=false)
- s/extra/container
- added keyword args to magic __ methods
- added documentation
- disallow multiple docker command
- make docker_command and container_command not cumulative

hannes-ucsc commented:
@heuerm, I looked at all docker command `--help`'s and found only three commands that have these "extra" arguments: `create`, `run` and `exec`. So I renamed `extra` to `container`. What do you think?

heuermh commented:
Yes, I like `container` better than `extra`.  Thanks for thinking on it!
